File: crates/html-to-markdown/src/options.rs
//! Configuration options for HTML to Markdown conversion.

/// Heading style options.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HeadingStyle {
    /// Underlined style (=== for h1, --- for h2)
    Underlined,
    /// ATX style (# for h1, ## for h2, etc.)
    Atx,
    /// ATX closed style (# title #)
    AtxClosed,
}

impl Default for HeadingStyle {
    fn default() -> Self {
        Self::Underlined
    }
}

/// List indentation type.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ListIndentType {
    Spaces,
    Tabs,
}

impl Default for ListIndentType {
    fn default() -> Self {
        Self::Spaces
    }
}

/// Whitespace handling mode.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum WhitespaceMode {
    Normalized,
    Strict,
}

impl Default for WhitespaceMode {
    fn default() -> Self {
        Self::Normalized
    }
}

/// Newline style.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum NewlineStyle {
    /// Two spaces at end of line
    Spaces,
    /// Backslash at end of line
    Backslash,
}

impl Default for NewlineStyle {
    fn default() -> Self {
        Self::Spaces
    }
}

/// Highlight style for `<mark>` elements.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HighlightStyle {
    /// ==text==
    DoubleEqual,
    /// <mark>text</mark>
    Html,
    /// **text**
    Bold,
    /// Plain text (no formatting)
    None,
}

impl Default for HighlightStyle {
    fn default() -> Self {
        Self::DoubleEqual
    }
}

/// Preprocessing preset levels.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PreprocessingPreset {
    Minimal,
    Standard,
    Aggressive,
}

impl Default for PreprocessingPreset {
    fn default() -> Self {
        Self::Standard
    }
}

/// Main conversion options.
#[derive(Debug, Clone)]
pub struct ConversionOptions {
    /// Heading style
    pub heading_style: HeadingStyle,

    /// List indentation type
    pub list_indent_type: ListIndentType,

    /// List indentation width (spaces)
    pub list_indent_width: usize,

    /// Bullet characters for unordered lists
    pub bullets: String,

    /// Symbol for strong/emphasis (* or _)
    pub strong_em_symbol: char,

    /// Escape asterisks in text
    pub escape_asterisks: bool,

    /// Escape underscores in text
    pub escape_underscores: bool,

    /// Escape misc markdown characters
    pub escape_misc: bool,

    /// Default code language
    pub code_language: String,

    /// Use autolinks for bare URLs
    pub autolinks: bool,

    /// Add default title if none exists
    pub default_title: bool,

    /// Use <br> in tables instead of spaces
    pub br_in_tables: bool,

    /// Highlight style for <mark> elements
    pub highlight_style: HighlightStyle,

    /// Extract metadata from HTML
    pub extract_metadata: bool,

    /// Whitespace handling mode
    pub whitespace_mode: WhitespaceMode,

    /// Strip newlines from HTML before processing
    pub strip_newlines: bool,

    /// Enable text wrapping
    pub wrap: bool,

    /// Text wrap width
    pub wrap_width: usize,

    /// Treat block elements as inline
    pub convert_as_inline: bool,

    /// Subscript symbol
    pub sub_symbol: String,

    /// Superscript symbol
    pub sup_symbol: String,

    /// Newline style
    pub newline_style: NewlineStyle,

    /// Elements where images should remain as markdown (not converted to alt text)
    pub keep_inline_images_in: Vec<String>,

    /// Preprocessing options
    pub preprocessing: PreprocessingOptions,

    /// Parsing options
    pub parsing: ParsingOptions,
}

impl Default for ConversionOptions {
    fn default() -> Self {
        Self {
            heading_style: HeadingStyle::default(),
            list_indent_type: ListIndentType::default(),
            list_indent_width: 4,
            bullets: "*+-".to_string(),
            strong_em_symbol: '*',
            escape_asterisks: true,
            escape_underscores: true,
            escape_misc: true,
            code_language: String::new(),
            autolinks: true,
            default_title: false,
            br_in_tables: false,
            highlight_style: HighlightStyle::default(),
            extract_metadata: true,
            whitespace_mode: WhitespaceMode::default(),
            strip_newlines: false,
            wrap: false,
            wrap_width: 80,
            convert_as_inline: false,
            sub_symbol: String::new(),
            sup_symbol: String::new(),
            newline_style: NewlineStyle::default(),
            keep_inline_images_in: Vec::new(),
            preprocessing: PreprocessingOptions::default(),
            parsing: ParsingOptions::default(),
        }
    }
}

/// HTML preprocessing options.
#[derive(Debug, Clone)]
pub struct PreprocessingOptions {
    /// Enable preprocessing
    pub enabled: bool,

    /// Preprocessing preset
    pub preset: PreprocessingPreset,

    /// Remove navigation elements
    pub remove_navigation: bool,

    /// Remove form elements
    pub remove_forms: bool,
}

impl Default for PreprocessingOptions {
    fn default() -> Self {
        Self {
            enabled: false,
            preset: PreprocessingPreset::default(),
            remove_navigation: true,
            remove_forms: true,
        }
    }
}

/// HTML parsing options.
#[derive(Debug, Clone)]
pub struct ParsingOptions {
    /// Source encoding
    pub encoding: String,

    /// HTML parser to use
    pub parser: Option<String>,
}

impl Default for ParsingOptions {
    fn default() -> Self {
        Self {
            encoding: "utf-8".to_string(),
            parser: None,
        }
    }
}

File: packages/python/html_to_markdown/options.py
"""Configuration options for HTML to Markdown conversion.

This module provides dataclass-based configuration for the v2 API.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Literal


class OutputFormat(str, Enum):
    """Output format for conversion.

    Specifies the target markup language format for the conversion output.

    Attributes:
        MARKDOWN: Standard Markdown (CommonMark compatible). Default.
        DJOT: Djot lightweight markup language.

    Example:
        >>> options = ConversionOptions(output_format=OutputFormat.DJOT)
        >>> # Or use string literal: output_format="djot"
    """

    MARKDOWN = "markdown"
    """Standard Markdown (CommonMark compatible). Default."""

    DJOT = "djot"
    """Djot lightweight markup language."""


@dataclass
class ConversionOptions:
    """Main conversion configuration.

    This class groups all conversion-related options together, replacing
    the large number of keyword arguments in the v1 API.

    Example:
        >>> options = ConversionOptions(
        ...     heading_style="atx",
        ...     list_indent_width=2,
        ...     escape_asterisks=True,
        ... )
        >>> from html_to_markdown import convert
        >>> markdown = convert("<h1>Title</h1>", options)
    """

    heading_style: Literal["underlined", "atx", "atx_closed"] = "atx"
    """Style for headings: 'atx' (#) is CommonMark default, 'underlined' (===), or 'atx_closed' (# #)."""

    list_indent_type: Literal["spaces", "tabs"] = "spaces"
    """Type of indentation for lists."""

    list_indent_width: int = 2
    """Number of spaces for list indentation (CommonMark uses 2 spaces, ignored if list_indent_type='tabs')."""

    bullets: str = "-*+"
    """Characters to use for unordered list bullets (cycles through -, *, + for nested levels). CommonMark compliant."""

    strong_em_symbol: Literal["*", "_"] = "*"
    """Symbol for strong/emphasis formatting."""

    escape_asterisks: bool = False
    """Escape asterisk characters in text to prevent accidental formatting. Default False for minimal escaping (CommonMark)."""

    escape_underscores: bool = False
    """Escape underscore characters in text to prevent accidental formatting. Default False for minimal escaping (CommonMark)."""

    escape_misc: bool = False
    """Escape miscellaneous Markdown characters. Default False for minimal escaping (CommonMark)."""

    escape_ascii: bool = False
    """Escape all ASCII punctuation (for CommonMark spec compliance tests). Disabled by default for minimal escaping."""

    code_language: str = ""
    """Default language for code blocks."""

    encoding: str = "utf-8"
    """Character encoding expected for the HTML input."""

    autolinks: bool = True
    """Convert bare URLs to automatic links."""

    default_title: bool = False
    """Add a default title if none exists."""

    keep_inline_images_in: set[str] | None = None
    """Parent tag names where images should remain inline."""

    br_in_tables: bool = False
    """Use <br> tags for line breaks in table cells instead of spaces."""

    highlight_style: Literal["double-equal", "html", "bold"] = "double-equal"
    """Style for highlighting <mark> elements."""

    extract_metadata: bool = True
    """Extract metadata from HTML head and include as comment."""

    whitespace_mode: Literal["normalized", "strict"] = "normalized"
    """How to handle whitespace: 'normalized' or 'strict'."""

    strip_newlines: bool = False
    """Remove newlines from HTML before processing."""

    wrap: bool = False
    """Enable text wrapping."""

    wrap_width: int = 80
    """Column width for text wrapping."""

    strip_tags: set[str] | None = None
    """HTML tags to strip from output (output only text content, no markdown conversion)."""

    preserve_tags: set[str] | None = None
    """HTML tags to preserve as-is in the output (keep original HTML). Useful for complex elements like tables."""

    skip_images: bool = False
    """Skip all images during conversion.

    When enabled, all `<img>` elements are completely omitted from output.
    Useful for text-only extraction or filtering out visual content.
    """

    convert_as_inline: bool = False
    """Treat block elements as inline during conversion."""

    sub_symbol: str = ""
    """Symbol for subscript text."""

    sup_symbol: str = ""
    """Symbol for superscript text."""

    newline_style: Literal["spaces", "backslash"] = "spaces"
    """Style for newlines: 'spaces' (two trailing spaces, CommonMark default) or 'backslash' (\\). Both are equally CommonMark compliant."""

    code_block_style: Literal["indented", "backticks", "tildes"] = "backticks"
    """Style for code blocks: 'backticks' (```, better whitespace preservation), 'indented' (4 spaces), or 'tildes' (~~~). All are CommonMark compliant."""

    output_format: Literal["markdown", "djot"] = "markdown"
    """Output format: 'markdown' (CommonMark, default) or 'djot' (lightweight markup language)."""

    debug: bool = False
    """Enable debug mode with diagnostic warnings about unhandled elements and hOCR processing."""

    include_document_structure: bool = False
    """Include document structure in conversion output."""

    extract_images: bool = False
    """Extract images during conversion."""

    max_image_size: int = 5_242_880
    """Maximum image size in bytes (default 5 MiB)."""

    capture_svg: bool = False
    """Capture SVG elements as images."""

    infer_dimensions: bool = True
    """Infer image dimensions from data URIs."""

    max_depth: int = 100
    """Maximum DOM tree depth to recurse into. Nodes beyond this depth are silently skipped."""


@dataclass
class PreprocessingOptions:
    """HTML preprocessing configuration.

    Controls how HTML is cleaned and preprocessed before conversion.

    Example:
        >>> options = PreprocessingOptions(
        ...     enabled=True,
        ...     preset="aggressive",
        ...     remove_navigation=True,
        ... )
    """

    enabled: bool = True
    """Whether to enable HTML preprocessing (enabled by default for robust handling of malformed HTML)."""

    preset: Literal["minimal", "standard", "aggressive"] = "standard"
    """Preprocessing aggressiveness level."""

    remove_navigation: bool = True
    """Remove navigation elements during preprocessing."""

    remove_forms: bool = True
    """Remove form elements during preprocessing."""
